📖clusterctl: doc config for air-gapped environments (#2595)
kubernetes-sigs/cluster-api: docs/book/src/clusterctl/configuration.md

```

In case a variable is defined both in the config file and as an OS environment variable, the latter takes precedence.

## Image overrides

<aside class="note warning">

<h1> Warning! </h1>

Image override is an advanced feature and wrong configuration can easily lead to non-functional clusters.
It's strongly recommended to test configurations on dev/test environments before using this functionality in production.

</aside>

When working in air-gapped environments, it's necessary to alter the manifests to be installed in order to pull
images from a local/custom image repository instead of public ones (e.g. `gcr.io`, or `quay.io`).

The `clusterctl` configuration file can be used to instruct `clusterctl` to override images automatically.

This can be achieved by adding an `images` configuration entry as shown in the example:

```yaml
images:
  all:
    repository: myorg.io/local-repo
```

Please note that the image override feature allows for more fine-grained configuration, allowing to set image
overrides for specific components, for example:

```yaml
images:
  all:
    repository: myorg.io/local-repo

>   cert-manager:

Reviewer: can all component images be overridden? If not, would it make sense to provide a list of images that can be?

Author: @CecileRobertMichon 
ATM we are not supporting overrides for a single image, but only overrides at component level. In this example, "cert-manager" means all the images in the cert-manager manifest.

Available components are "cert-manger", "cluster-api" and all the installed providers: "bootstrap-kubeadm", "control-plane-kubeadm", "infrastracture-aws" etc.

Eventually, in the future/if there are specific needs, the mechanism can be extended for implementing image specific settings, but at the end the list of images depends on the provider manifests so it will be up to the provides to document which images are used.

>     tag: v0.11.1 

Reviewer: Will the image specific override take precedence over the `all` setting?

Author: @CecileRobertMichon 
The override works union, with the most specific override config taking precedence on the all config in case of conflicts.

So, in this example:
-  for cluster-api images, the "all" override apply to all the images in the cluster-api manifest; same for other providers
-  for cert-manager images, the "all" override U the "cert-manager" override apply to all the images in the cert-manager manifest

```

In case a variable is defined both in the config file and as an OS environment variable, the latter takes precedence.

## Image overrides

> When working on air-gapped environments it is necessary to alter the manifest to be installed in order to pull

Reviewer 2: ```suggestion
When working in air-gapped environments, it's necessary to alter the manifests to be installed in order to pull
```

> images from a local image repository instead of pulling from public image repositories like gcr.io or quay.io.

Reviewer 2: ```suggestion
images from a local/custom image repository instead of public ones (e.g. `gcr.io`, or `quay.io`).
```

The `clusterctl` config file can be used to instruct `clusterctl` to execute automatic image repository override, instead of

> changing YAML files manually.

Reviewer 2: ```suggestion
The `clusterctl` configuration file can be used to instruct `clusterctl` to override images automatically.
```

This can be achieved by adding an `images` configuration entry as shown in the example:

> ```

Reviewer 2: ```suggestion
```yaml
```

images:
  all:
    repository: myorg.io/local-repo
```

Please note that the image override feature allows for more fine-grained configuration, allowing to set image

> ovverrides for specific components e.g.

Reviewer 2: ```suggestion
overrides for specific components, for example:
```

> ```

Reviewer 2: ```suggestion
```yaml
```

images:
  all:
    repository: myorg.io/local-repo
  cert-manager:
    tag: v0.11.1 
```

In this example we are overriding the image repository for all the components and the image tag for
all the images in the cert-manager component.

> <aside class="note warning">

Reviewer 2: This should probably be at the top of this section

<h1> Warning! </h1>

> Image override is an advanced feature and wrong configuration can easily lead to not working management clusters,

Reviewer 2: ```suggestion
Image override is an advanced feature and wrong configuration can easily lead to non-functional clusters.
```

> so it is strongly recommended to test configurations on dev/test environments before applying on production.

Reviewer 2: ```suggestion
It's strongly recommended to test configurations on dev/test environments before using this functionality in production.
```

E.g. if you set the image tag for a provider components, e.g. the `infrastrcutrue-aws` provider, this most probably will lead to a not working management cluster because `kube-rbac-proxy` and

> `cluster-api-aws-controller` usually are expected to have two different image tags.

Reviewer 2: ```suggestion
Example: Let's say we're working with `infrastructure-aws`, overriding the image tag in the default provider components 
is going to result in a non-functional cluster due to `kube-rbac-proxy` and `cluster-api-aws-controller` having different tags.
```

Author: I have dropped the example because it is hard to understand after moving the warning at the beginning of the paragraph
